File: callNOAA.py
import pandas as pd
import json
import numpy as np
from datetime import datetime
import requests


import matplotlib.pyplot as plt
import plotly.express as px
import plotly.io as pio


#access token noaa
Token = 'JsDVdrTYMcumWUloSrfsYLQFneUVwnie'

#initialize lists to store data
dates_temp = []
dates_prcp = []
temps = []
prcp = []

#for each year from 2015-2019 ...
for year in range(2015, 2020):
    year = str(year)
    r = requests.get('https://www.ncdc.noaa.gov/cdo-web/api/v2/data?datasetid=GHCND&datatypeid=TAVG&limit=1000&stationid=GHCND:USW00013743&startdate='+year+'-01-01&enddate='+year+'-12-31', headers={'token':Token})
    d = json.loads(r.text)
    avg_temps = [item for item in d['results'] if item['datatype']=='TAVG']
    dates_temp += [item['date'] for item in avg_temps]
    temps += [item['value'] for item in avg_temps]

df_temp = pd.DataFrame()

df_temp['date'] = [datetime.strptime(d, "%Y-%m-%dT%H:%M:%S") for d in dates_temp]
df_temp['avgTemp'] = [float(v)/10.0*1.8 + 32 for v in temps]
# df_temp.plot(x ='date', y='avgTemp', kind = 'line')
# plt.show()

fig = px.line(df_temp, x='date', y='avgTemp')
pio.write_html(fig, file='indexw.html', auto_open=True)

File: callNOAA.js
var weatherData = {}; //data from both stations from request to NOAA API
var stationName1, stationName2;
var stationNameToId = {}; //map name to GHCND ID
var APIkey = 'JsDVdrTYMcumWUloSrfsYLQFneUVwnie'; // for NOAA API, read from external file


var periodMode, startdate, enddate;

function callNOAA(month,year,stationID) {
  if(month=="ALL MONTHS"){
  	periodMode = "year";
  	startDate = year+"-01-01";
  	endDate = year+"-12-31";
  }
  else{
  	periodMode = "month";
  	var monthCode = monthToMonth(month);
  	startDate = year+"-"+monthCode+"-01";
  	if(monthCode == '01' || monthCode == '03' || monthCode == '05' || monthCode == '07' || monthCode == '08'|| monthCode == '10' || monthCode == '12'){
  		endDate = year+"-"+monthCode+"-31";
  	}
  	else if(monthCode=='02'){
  		var numYr = parseInt(year);
  		if(numYr%4==0 && !(numYr%100==0)){
  			endDate = year+"-"+monthCode+"-29";
  		}
  		else if(numYr%100==0 && numYr%400==0){
  			endDate = year+"-"+monthCode+"-30";
  		}
  		else{
  			endDate = year+"-"+monthCode+"-28";
  		}
  	}
  	else{
  		endDate = year+"-"+monthCode+"-30";
  	}
  	
  }
//stationID = 'US1MDBC0003';
console.log(stationID);
var url = 'https://www.ncdc.noaa.gov/cdo-web/api/v2/data?datasetid=GHCND&datatypeid=TAVG&limit=1000&stationid=GHCND:'+stationID+'&startdate='+startDate+'&enddate='+endDate;
var responseData = makeHttpRequest(url, APIkey);
console.log(responseData);
  if (Object.keys(responseData).length == 0) {
    alert('No NOAA data for ' + 'this' + ' station for those dates');
    document.getElementById("no-selection-avg-temp").innerHTML = "<i>Select a station, month, and year to view graph.</i>";
  }
  return responseData;
}

function makeHttpRequest(url, APIkey) {
  var xhttp = new XMLHttpRequest();
  xhttp.onreadystatechange = function() {
    if (xhttp.readyState == 4 && xhttp.status == 200) { // readyState == DONE and status == OK
      //document.getElementById("debug").innerHTML = xhttp.responseText; //for debugging
      responseData = JSON.parse(xhttp.responseText);
    }
  };
  xhttp.open("GET", url, false);
  xhttp.setRequestHeader("token", APIkey.trim());
  xhttp.send();
  return responseData;
}

function range(start, end) {
    if(start === end) return [start];
    return [start, ...range(start + 1, end)];
}

function monthToMonth(month){
	switch(month) {
	    case 'January':
	        return('01');
	    case 'February':
	        return('02');
	    case 'March':
	        return('03');
	    case 'April':
	        return('04');
	    case 'May':
	        return('05');
	    case 'June':
	        return('06');
	    case 'July':
	        return('07');
	    case 'August':
	        return('08');
	    case 'September':
	        return('09');
	    case 'October':
	        return('10');
	    case 'November':
	        return('11');
	    case 'December':
	        return('12');

	}
}


